### ISSUE TITLE
Make scaler.step() return if step was skipped or not

### ISSUE THREAD

**Issue by pyphan1** [pytorch/pytorch#152279]
### 🚀 The feature, motivation and pitch

Make calling scaler.step(optimizer) return if the step was skipped or not instead of always returning None, or make it print when it skips a step
for example we can use:
stepped = scaler.step(optimizer)
if not stepped: 
    print('Step was skipped because of an underflow or overflow')

it is easy to implement returning the stepping state, it should be returned from grad_scaler._maybe_opt_step

for example, some models apply lazy regularization or lazy loss types that are expensive to compute, imagine that you're training a model with an expensive loss type that you apply every 16 steps, you also multiply the loss by 16 to make it as strong as if it is applied in every step, now imagine that your loss grew 4x in the step that you add this expensive loss type in, after scaling the loss with the scaler that used to scale normal loss you might get NaNs/infs in the loss due to an overflow/underflow, and the step will be skipped automatically, however you will never know and this can happen every time you add the lazy loss, so it will effectively be like you never used this necessary loss function.

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by AbhiLegend** [comment]
Fix scaled_dot_product_attention (SDPA) slowdown by disabling memory-efficient kernel to avoid CUDA Graph capture context switching.
Steps Taken
1. Set PyTorch SDPA backend flags at the beginning of the program:
  `import torch.backends.cuda.sdp_kernel
torch.backends.cuda.sdp_kernel.enable_math = False
torch.backends.cuda.sdp_kernel.enable_flash = True
torch.backends.cuda.sdp_kernel.enable_mem_efficient = False
`
2. Ensure that scaled_dot_product_attention only uses FlashAttention kernels.

3.Prevent context switching between normal execution and CUDA graph captured streams, which can cause major slowdowns.

4.Validated performance improvement by timing 10 forward passes before and after setting the flags.

I am attaching the notebook [](https://colab.research.google.com/drive/1E6ikc37LLKYsQo6eycm15RPuThdEMy_1?usp=sharing) .If this looks good assign the issue to me  @pyphan1

**Comment by pyphan1** [comment]
@AbhiLegend i believe you replied to the wrong issue

**Comment by janeyx99** [comment]
Hi @pyphan1, the `_maybe_opt_step` function already has a return value--it returns whatever the step returns:
```
def _maybe_opt_step(
        self,
        optimizer: torch.optim.Optimizer,
        optimizer_state: dict[str, Any],
        *args: Any,
        **kwargs: Any,
    ) -> Optional[float]:
        retval: Optional[float] = None
        if not sum(v.item() for v in optimizer_state["found_inf_per_device"].values()):
            retval = optimizer.step(*args, **kwargs)
        return retval
```

As we've already defined the semantic for this function (and therefore `scaler.step`), it will be backwards incompatible to change the semantic. However! An easy way to get the behavior that you want is by passing in a closure into step that will always return a non-None value (like True). This way, if step doesn't get called, you would get a None, and if it does, these functions will both return True! This is definitely a pretty hacky way to use the closure...but it should unblock lol.

Example pseudocode:

```
optim = torch.optim.AdamW(model.named_parameters(), fused=True)
... // fwd, bwd, etc
optim.step(lambda: True)
```

**Comment by pyphan1** [comment]
@janeyx99 Thank you!
